File: android/app/src/main/java/com/h19h29/naymnaymlevelup/rebuild/child/ForestSceneMotion.kt
package com.h19h29.naymnaymlevelup.rebuild.child

enum class ForestSceneLayer(
    val zIndex: Float,
) {
    Sky(0f),
    DistantTrees(1f),
    MidgroundTrees(2f),
    ForegroundLeaves(3f),
    Ground(4f),
    ;

    companion object {
        const val ContentZIndex = 10f
    }
}

data class ForestSceneOffset(
    val x: Float,
    val y: Float,
) {
    companion object {
        val Zero = ForestSceneOffset(0f, 0f)
    }
}

data class ForestSceneFrame(
    val progress: Float,
    private val transforms: Map<ForestSceneLayer, ForestSceneOffset>,
) {
    operator fun get(layer: ForestSceneLayer): ForestSceneOffset =
        transforms[layer] ?: ForestSceneOffset.Zero
}

data class ForestSceneActivity(
    val isSheetPresented: Boolean,
    val isTabActive: Boolean,
    val isAppActive: Boolean,
) {
    val isPaused: Boolean
        get() = isSheetPresented || !isTabActive || !isAppActive

    companion object {
        val Active = ForestSceneActivity(
            isSheetPresented = false,
            isTabActive = true,
            isAppActive = true,
        )
    }
}

class ForestSceneMotionClock(
    private val startNanos: Long,
) {
    private var accumulatedPauseNanos = 0L
    private var pauseStartedAtNanos: Long? = null

    fun update(
        activity: ForestSceneActivity,
        atNanos: Long,
    ) {
        if (activity.isPaused) {
            if (pauseStartedAtNanos == null) {
                pauseStartedAtNanos = atNanos
            }
        } else {
            pauseStartedAtNanos?.let { pauseStart ->
                accumulatedPauseNanos += (atNanos - pauseStart).coerceAtLeast(0L)
                pauseStartedAtNanos = null
            }
        }
    }

    fun elapsedMillis(atNanos: Long): Long {
        val effectiveNanos = pauseStartedAtNanos ?: atNanos
        return (
            effectiveNanos - startNanos - accumulatedPauseNanos
        ).coerceAtLeast(0L) / 1_000_000L
    }
}

object ForestSceneMotionSpec {
    const val CycleDurationMillis = 8_000L

    fun frameAt(
        elapsedMillis: Long,
        reduceMotion: Boolean,
    ): ForestSceneFrame {
        if (reduceMotion) {
            return ForestSceneFrame(
                progress = 0f,
                transforms = ForestSceneLayer.entries.associateWith {
                    ForestSceneOffset.Zero
                },
            )
        }
        val wrapped = elapsedMillis.coerceAtLeast(0L) % CycleDurationMillis
        val halfCycle = CycleDurationMillis / 2f
        val progress = if (wrapped <= CycleDurationMillis / 2) {
            wrapped / halfCycle
        } else {
            (CycleDurationMillis - wrapped) / halfCycle
        }
        return ForestSceneFrame(
            progress = progress,
            transforms = mapOf(
                ForestSceneLayer.Sky to ForestSceneOffset.Zero,
                ForestSceneLayer.DistantTrees to ForestSceneOffset(
                    x = 0f,
                    y = -2f * progress,
                ),
                ForestSceneLayer.MidgroundTrees to ForestSceneOffset(
                    x = 0f,
                    y = -4f * progress,
                ),
                ForestSceneLayer.ForegroundLeaves to ForestSceneOffset(
                    x = 6f * progress,
                    y = -3f * progress,
                ),
                ForestSceneLayer.Ground to ForestSceneOffset.Zero,
            ),
        )
    }

    fun shouldScheduleFrameCallback(
        reduceMotion: Boolean,
        activity: ForestSceneActivity,
    ): Boolean = !reduceMotion && !activity.isPaused
}

File: NaymNaymLevelUp/Rebuild/Child/ForestSceneMotion.swift
import CoreGraphics
import Foundation

enum ForestSceneLayer: String, CaseIterable, Equatable {
    case sky
    case distantTrees
    case midgroundTrees
    case foregroundLeaves
    case ground

    static let contentZIndex = 10.0

    var zIndex: Double {
        switch self {
        case .sky: return 0
        case .distantTrees: return 1
        case .midgroundTrees: return 2
        case .foregroundLeaves: return 3
        case .ground: return 4
        }
    }
}

struct ForestSceneOffset: Equatable {
    var x: CGFloat
    var y: CGFloat

    static let zero = ForestSceneOffset(x: 0, y: 0)
}

struct ForestSceneFrame: Equatable {
    let progress: CGFloat
    private let transforms: [ForestSceneLayer: ForestSceneOffset]

    init(
        progress: CGFloat,
        transforms: [ForestSceneLayer: ForestSceneOffset]
    ) {
        self.progress = progress
        self.transforms = transforms
    }

    subscript(layer: ForestSceneLayer) -> ForestSceneOffset {
        transforms[layer] ?? .zero
    }
}

struct ForestSceneActivity: Equatable {
    let isSheetPresented: Bool
    let isTabActive: Bool
    let isAppActive: Bool

    static let active = ForestSceneActivity(
        isSheetPresented: false,
        isTabActive: true,
        isAppActive: true
    )

    var isPaused: Bool {
        isSheetPresented || !isTabActive || !isAppActive
    }
}

struct ForestSceneMotionClock {
    private let startTime: TimeInterval
    private var accumulatedPauseDuration: TimeInterval = 0
    private var pauseStartedAt: TimeInterval?

    init(startTime: TimeInterval) {
        self.startTime = startTime
    }

    mutating func update(
        activity: ForestSceneActivity,
        at time: TimeInterval
    ) {
        if activity.isPaused {
            if pauseStartedAt == nil {
                pauseStartedAt = time
            }
        } else if let pauseStartedAt {
            accumulatedPauseDuration += max(0, time - pauseStartedAt)
            self.pauseStartedAt = nil
        }
    }

    func elapsed(at time: TimeInterval) -> TimeInterval {
        let effectiveTime = pauseStartedAt ?? time
        return max(0, effectiveTime - startTime - accumulatedPauseDuration)
    }
}

enum ForestSceneMotionSpec {
    static let cycleDuration: TimeInterval = 8

    static func frame(
        elapsed: TimeInterval,
        reduceMotion: Bool
    ) -> ForestSceneFrame {
        guard !reduceMotion else {
            return ForestSceneFrame(
                progress: 0,
                transforms: Dictionary(
                    uniqueKeysWithValues: ForestSceneLayer.allCases.map {
                        ($0, .zero)
                    }
                )
            )
        }

        let wrapped = max(0, elapsed).truncatingRemainder(
            dividingBy: cycleDuration
        )
        let progress = CGFloat(
            wrapped <= cycleDuration / 2
                ? wrapped / (cycleDuration / 2)
                : (cycleDuration - wrapped) / (cycleDuration / 2)
        )
        return ForestSceneFrame(
            progress: progress,
            transforms: [
                .sky: .zero,
                .distantTrees: ForestSceneOffset(x: 0, y: -2 * progress),
                .midgroundTrees: ForestSceneOffset(x: 0, y: -4 * progress),
                .foregroundLeaves: ForestSceneOffset(
                    x: 6 * progress,
                    y: -3 * progress
                ),
                .ground: .zero,
            ]
        )
    }

    static func shouldScheduleFrameCallback(
        reduceMotion: Bool,
        activity: ForestSceneActivity
    ) -> Bool {
        !reduceMotion && !activity.isPaused
    }
}
